contracts/tokens/SecurityToken.sol

```diff
@@ -394,6 +411,143 @@ contract SecurityToken is ERC20, ERC20Detailed, Ownable, ReentrancyGuard, Securi
+        return _transferByPartition(msg.sender, _to, _value, _partition, _data, address(0), "");
+    }
+
+    function _balanceOfByPartition(bytes32 _partition, address _tokenHolder, uint256 _additionalBalance) internal view returns(uint256) {
+        address[] memory tms = modules[TRANSFER_KEY];
+        uint256 max;
+        uint256 amount;
+        for (uint256 i = 0; i < tms.length; i++) {
+            amount = ITransferManager(tms[i]).getTokensByPartition(_partition, _tokenHolder, _additionalBalance);
+            if (max < amount) {
+                max = amount;
+            }
+        }
+        return max;
+    }
+
+    function _transferByPartition(
+        address _from,
+        address _to,
+        uint256 _value,
+        bytes32 _partition,
+        bytes memory _data,
+        address _operator,
+        bytes memory _operatorData
+    ) 
+        internal
+        returns(bytes32 toPartition) 
+    {
+        _isValidPartition(_partition);
+        // Avoiding to add this check 
+        // require(balanceOfByPartition(_partition, msg.sender) >= _value);
+        // NB - Above condition will be automatically checked using the executeTransfer() function execution.
+        // NB - passing `_additionalBalance` value is 0 because accessing the balance before transfer
+        uint256 balanceBeforeTransferLocked = _balanceOfByPartition(LOCKED, _to, 0);
+        _transferWithData(_from, _to, _value, _data);
+        // NB - passing `_additonalBalance` valie is 0 because balance of `_to` was updated in the transfer call
+        uint256 balanceAfterTransferLocked = _balanceOfByPartition(LOCKED, _to, 0);
+        toPartition =  _returnPartition(balanceBeforeTransferLocked, balanceAfterTransferLocked, _value);
+        emit TransferByPartition(toPartition, _operator, _from, _to, _value, _data, _operatorData);
+    }
+
+    function _returnPartition(uint256 _beforeBalance, uint256 _afterBalance, uint256 _value) internal pure returns(bytes32 toPartition) {
+        // return LOCKED only when the transaction `_value` should be equal to the change in the LOCKED partition
+        // balance otherwise return UNLOCKED
+        if (_afterBalance.sub(_beforeBalance) == _value)
+            toPartition = LOCKED;
+        // Returning the same partition UNLOCKED 
+        toPartition = UNLOCKED;
+    }
+
+    ///////////////////////
+    /// Operator Management
+    ///////////////////////
+    
+    /**
+     * @notice Authorises an operator for all partitions of `msg.sender`.
+     * NB - Allowing investors to authorize an investor to be an operator of all partitions
+     * but it doesn't mean we operator is allowed to transfer the LOCKED partition values.
+     * Logic for this restriction is written in `operatorTransferByPartition()` function.
+     * @param _operator An address which is being authorised.
+     */ 
+    function authorizeOperator(address _operator) external {
+        approvals[msg.sender][_operator] = true;
```

Comment:
Possible alternative approach I've seen which could maybe avoid having a separate `approvals` value is to use the normal ERC20 `allowances` variable, but set the allowance to uint(-1) and then don't subtract the amount in transferFrom if the allowance is uint(-1).

WETH uses this approach:
https://etherscan.io/address/0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2#code

Reply:
Yes, it will work with `authorizeOperator`. That wouldn't work with `authorizeOperatorByPartition` in general but for our current case, it will work as we only have UNLOCKED partition to entertain. But if we want to introduce more tradable partition in the future then upgradeability will not help as we need to introduce more state variables to tackle this. So I will use uint(-1) approach for the `approvals` not for the `partitionApprovals`